mesonbuild/scripts/run_tool.py

```diff
@@ -3,17 +3,85 @@
 
 from __future__ import annotations
 
-import itertools
+import asyncio
+import asyncio.subprocess
 import fnmatch
-import concurrent.futures
+import signal
+import sys
 from pathlib import Path
 
+from .. import mlog
 from ..compilers import lang_suffixes
 from ..mesonlib import quiet_git
+from ..mtest import sh_quote, determine_worker_count, complete_all
 import typing as T
 
-if T.TYPE_CHECKING:
-    import subprocess
+Info = T.TypeVar("Info")
+
+async def run_with_buffered_output(cmdlist: T.List[str]) -> int:
+    """Run the command in cmdlist, buffering the output so that it is
+       not mixed for multiple child processes.  Kill the child on
+       cancellation."""
+    p = None
+    quoted_cmdline = ' '.join(sh_quote(x) for x in cmdlist)
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
+        stdo, _ = await p.communicate()
+        if stdo:
+            print(mlog.blue('>>>'), quoted_cmdline)
+            sys.stdout.flush()
+            sys.stdout.buffer.write(stdo)
+        return p.returncode
+    except FileNotFoundError as e:
+        print(mlog.blue('>>>'), quoted_cmdline, file=sys.stderr)
+        print(mlog.red('not found:'), e.filename, file=sys.stderr)
+        return 1
+    except asyncio.CancelledError:
+        if p:
+            p.kill()
+            await p.wait()
+            return p.returncode or 1
+        else:
+            return 0
+
+async def _run_workers(infos: T.Iterable[Info],
+                       fn: T.Callable[[Info], T.Iterable[T.Coroutine[T.Any, T.Any, int]]]) -> int:
```

Comment:
I think you can simplify this to `fn: T.Callable[[Info], T.Iterable[T.Awaitable[int]]]`

Bonus points that it remove the need for `Any`

Reply:
Unfortunately you need Coroutine to be able to throw a CancellableError into it. Without the `throw`, you get a "coroutine was never awaited" error.

Reply:
In that case I think it should be `T.Coroutine[None, None, int]`, since the documentation says the annotations are the same as `Generator`; https://docs.python.org/3/library/typing.html#annotating-generators-and-coroutines